### PR TITLE
Add possibility to create functions without trailing semicolon

### LINKED ISSUE #7: Parse trailing semicolon in function body
It may be of significance to indicate what the function returns

```
fn test() -> Bool {
  true;
}
```

This feels like it doesn't return true, if you compare to this:

```
fn test() -> Bool {
  true
}
```

If we do think that's materially different, we should capture the presence of the last semicolon. Perhaps with an `Empty` expression or something.

### LINKED ISSUE #9: Allow for trailing expressions in code block
Right now, only this is possible:

```
{
return expr;
}
```

This has drawbacks, such as:

* It only commands a control flow of returning from a function (unless we introduce a `break` keyword that would break out of scopes)
* It's verbose

It would be nice to support this

```
{
 // ...;
expr
}
```

Notice that there's no trailing semicolon. That would be an implicit return value of a block. @p0lunin and I wanted to make it happen in #8 but couldn't quite make it right so we deferred this improvement until a later point.

### PR DESCRIPTION
**Pull request by p0lunin**
Closes #7. Closes #9.

### CODE REVIEW

File: lib/lexer.mll [hunk 1]
```diff
@@ -38,7 +38,6 @@ rule token = parse
  | '~' { TILDE }
  | '.' { DOT }
  | "let" { LET }
- | "type" { TYPE }
```
Comment (cmxog) [inline lib/lexer.mll:41]: Are we removing its "reserved" status here in this PR? I suppressed the unused token warning here 33ccbc6d4865ff56acea1570e6f1a25b8a4801d0

File: lib/syntax.ml [hunk 2]
```diff
@@ -141,16 +142,18 @@ functor
...
+    and code_block =
+      { block_exprs : expr located list; [@sexp.list]
+        return_value : expr located option [@sexp.option] }
 
     and if_ =
       { condition : expr located;
-        body : expr located list; [@sexp.list]
-        else_ : expr located option [@sexp.option] }
+        body : code_block located;
+        else_ : expr located option }
```
Comment (cmxog) [inline lib/syntax.ml]: any reason to drop `[@sexp.option]` here for `else_`? it hides `None` values from the sexp dumps `()`

File: lib/parser.mly [hunk 3]
```diff
@@ -160,27 +159,43 @@ let function_call :=
...
-let stmt :=
-  | terminated(expr, SEMICOLON)
-  | ~= terminated(let_binding, SEMICOLON); <Let>
+  | code_block = delimited(LBRACE, block_stmt, RBRACE);
+    { code_block }
+  | LBRACE; RBRACE; { make_code_block ~block_exprs:[] () }
+
+let block_stmt :=
+  | left = located(non_semicolon_stmt); right = block_stmt;
+    { make_code_block ~block_exprs:(left :: right.block_exprs) ?return_value:(right.return_value) () }
```
Comment (cmxog) [inline lib/parser.mly]: I wonder if we really need this `return_value` field. Can we construct `~block_exprs` to simply include `right.return_value` at the end?

File: lib/lang.ml [hunk 4]
```diff
@@ -417,6 +417,10 @@ functor
           function_returns = TermType return;
           function_body =
             Option.map f.exprs ~f:(fun exprs ->
+                let exprs =
+                  exprs.block_exprs
+                  @ match exprs.return_value with Some e -> [e] | None -> []
+                in
```
Comment (cmxog) [inline lib/lang.ml]: And if we do that transformation I mentioned at the parser level, we won't need to handle it in Lang.

**Comment by cmxog** [comment]
Overall, amazing progress. Left some commentary to make this nearly perfect.

File: lib/lexer.mll (hunk 1 above)
Comment (p0lunin) [reply lib/lexer.mll:41]: For now it's being removed until we satisfy our decision in https://github.com/tact-lang/spec/discussions/5

File: lib/lang.ml (hunk 4 above)
Comment (p0lunin) [reply lib/lang.ml]: Fixed in [6a9c1f6](https://github.com/tact-lang/tact/pull/18/commits/6a9c1f6b82356539639f365876deca023b2da018)

File: lib/parser.mly (hunk 3 above)
Comment (p0lunin) [reply lib/parser.mly]: Fixed in [6a9c1f6](https://github.com/tact-lang/tact/pull/18/commits/6a9c1f6b82356539639f365876deca023b2da018)

File: lib/syntax.ml (hunk 2 above)
Comment (p0lunin) [reply lib/syntax.ml]: Fixed in [6a9c1f6](https://github.com/tact-lang/tact/pull/18/commits/6a9c1f6b82356539639f365876deca023b2da018)

File: lib/lexer.mll (hunk 1 above)
Comment (cmxog) [reply lib/lexer.mll:41]: Ok

**Pull request merged**